pyface/testing/gui_test_case.py

```diff
@@ -0,0 +1,443 @@
+        ConditionTimeoutError
+            If the timeout occurs before the widget is destroyed.
+        """
+        destroy_later(control)
+        self.event_loop_until_control_destroyed(control, timeout)
+
+    def destroy_widget(self, widget, timeout=1.0):
+        """ Schedule a Widget for destruction and run the event loop.
+
+        Parameters
+        ----------
+        control : IWidget
+            The widget to destroy.
+        timeout : float
+            The number of seconds to run the event loop in the event that the
+            widget is not destroyed.
+
+        Raises
+        ------
+        ConditionTimeoutError
+            If the timeout occurs before the widget is destroyed.
+        """
+        if widget.control is not None:
+            control = widget.control
+            widget.destroy()
+            self.event_loop_until_control_destroyed(control, timeout)
+
+    # ------------------------------------------------------------------------
+    #  'TestCase' protocol
+    # ------------------------------------------------------------------------
+
+    def setUp(self):
+        """ Setup the test case for GUI interactions.
+        """
+        self.gui = GUI()
+        self.app = self.gui.app
+        self.event_loop_helper = EventLoopHelper(gui=self.gui)
+
+        if not os.environ.get('PYFACE_PATCH_ACTIVATE', False):
+            original_activated = Window.activate
+
+            def activate_patch(self, should_raise=True):
+                return original_activated(False)
+
+            self._raise_patch = mock.patch.object(
+                Window, 'activate', activate_patch)
+            self._raise_patch.start()
+        else:
+            self._raise_patch = None
+
+        self.gui.quit_on_last_window_close = False
+
+        # clean-up actions (LIFO)
+        self.addCleanup(self._delete_attrs, "gui", "app", "event_loop_helper")
+        self.addCleanup(self._restore_window_activate)
+        self.addCleanup(self._restore_quit_on_last_window_close)
+        self.addCleanup(self.gui.clear_event_queue)
+        self.addCleanup(self.gui.process_events)
+        self.addCleanup(self._close_top_level_windows)
+        self.addCleanup(self._gc_collect)
+        self.addCleanup(self.event_loop_helper.event_loop, 5, False)
+
+        super(GuiTestCase, self).setUp()
+
+    def tearDown(self):
+        """ Tear down the test case.
+
+        This method attempts to ensure that there are no windows that
+        remain open after the test case has run, and that there are no
+        events in the event queue to minimize the likelihood of one test
+        interfering with another.
+        """
+
+        super(GuiTestCase, self).tearDown()
+
+    def _gc_collect(self):
+        # Some top-level widgets may only be present due to cyclic garbage not
+        # having been collected; force a garbage collection before we decide to
+        # close windows. This may need several rounds.
+        for _ in range(10):
```

Comment:
It may be worth adding an "else" clause here that complains loudly (perhaps with a `warnings.warn` warning of some kind) if we reach it; if 10 rounds of garbage collection still haven't cleaned up everything, then something's likely seriously wrong (e.g., a separate thread that's busily creating new cycles while we're trying to clean them up) and worthy of investigation.